### ISSUE TITLE
question about thresholds

### ISSUE THREAD

**Issue by zilizili** [Tandon-A/emotic#2]
Hello, I would like to consult you. As mentioned in your code, threshold value can be used to help infer. I am not quite clear how this is used to infer.Looking forward to your reply.Thank you!

**Comment by Tandon-A** [comment]
@zilizili 
Hello, 

I apologize for the late reply. 
I am explaining the idea about using thresholds by giving an easier example first. 

Consider first, the case of having a problem where you need to classify an image in just one class, whether an image is of a car. In this case, you have just one neuron in the output layer with a sigmoid function. The sigmoid function outputs a number between [0,1] which is interpreted as the probability for the image being that of car. A 0.9 probability would mean a higher chance of the image being a car, while a 0.1 probability would mean the image is not a car. While the probability score is good to train the model you would also need a label at test time. How do you then convert this probability to a label? 
The idea is then to use a threshold to say whether the image belongs to the class or not. You would consider a sample as positively classified if the threshold is greater than this threshold. A good starting threshold with sigmoid activation is 0.5. But the value of this threshold depends on your problem and your performance metric. If this threshold is set to a high value like 0.85, then only when your model is very confident about the sample, it would be classified as a positive. This would decrease the false positives but increase the false negatives. In another case, if you set the threshold to a low value like 0.25, then your model would be able to classify all the cars in the dataset correctly. This would decrease the false negatives, but then increase the false positives, as many other objects would get classified as a car. 

Now let's consider the case of multi-class classification, where you need to classify an image in say three classes, 'car', 'cat' and 'house'. The output layer has three neurons outputting the probability score for the three classes. In this case, we say that an image belongs to only one of these classes. This constraint, allows us to use the SoftMax function, on the output layer which then converts the individual probabilities of the three neurons such that they all sum to one. The class which has the highest probability is then chosen as the label for the input image. The use of SoftMax thus removes the need for any thresholds. 

In our final case, we are taking the problem of emotion recognition, which is modelled as a multi-label classification problem.  For example, a person could be both 'happy' and 'excited'. Since an image can belong to multiple classes, we cannot use the SoftMax function now. Each neuron would now output a probability and we need to convert these probability scores to labels. Like the case of the single-class classification problem, we use thresholds to convert them to labels. 
Since we have 26 emotion classes, we have 26 threshold values to get the final labels. As mentioned, in the first example, a good starting point would be a threshold of 0.5 for all the classes. In this problem, I use the thresholds where precision is equal to recall for the class on the validation dataset. 
Check [this line](https://github.com/Tandon-A/emotic/blob/master/inference.py#L71) in the inference file where I compare the predictions with the thresholds. 

I hope this clarifies your doubt on how thresholds are used and gives you a good understanding of the problem as well. 

Let me know if you have any other concerns. 

Regards, 
Abhishek

**Comment by Tandon-A** [comment]
TL;DR :
For each class, the output neuron value is compared with the threshold for that class. If this value is greater than the threshold, then the image will have that class label.

**Comment by zilizili** [comment]
Thank you for your very detailed explanation, which is very helpful to me. Thanks again.
